presentation/src/main/java/com/pomonyang/presentation/base/BaseViewModel.kt

```diff
@@ -0,0 +1,77 @@
+package com.pomonyang.presentation.base
+
+import androidx.annotation.CallSuper
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+
+abstract class BaseViewModel<STATE : ViewState, EVENT : ViewEvent, EFFECT : ViewSideEffect> : ViewModel() {
+    // UI의 초기 상태를 설정
+    abstract fun setInitialState(): STATE
+
+    abstract suspend fun handleEvent(event: EVENT)
+
+    // 초기 상태를 지연 초기화
+    private val initialState: STATE by lazy { setInitialState() }
+
+    private val _state = MutableStateFlow<STATE>(initialState)
+    val state = _state.asStateFlow()
+
+    @Suppress("ktlint:standard:backing-property-naming")
+    private val _events: Channel<EVENT> = Channel<EVENT>(Channel.BUFFERED)
+
+    @Suppress("ktlint:standard:backing-property-naming")
+    private val _effects: Channel<EFFECT> = Channel<EFFECT>(Channel.BUFFERED)
+    val effect: Flow<EFFECT> = _effects.receiveAsFlow()
+
+    init {
+        subscribeToEvents()
+        onInit()
+    }
+
+    private fun subscribeToEvents() {
+        viewModelScope.launch {
+            for (event in _events) {
+                handleEvent(event)
+            }
+        }
+    }
+
+    // 추가적인 viewModel initial 작업
+    protected open fun onInit() {}
+
+    fun setEvent(event: EVENT) {
+        viewModelScope.launch { _events.send(event) }
+    }
+
+    protected fun setState(reducer: STATE.() -> STATE) {
+        val newState = _state.value.reducer()
+        _state.value = newState
```

Comment:
우리가 State를 data class로 사용을 할 것이고 State가 변경이 될 때마다 copy를 해야 하는 상황이라서 `value` 보다는 `update`로 기본 제공해주는게 좀 더 좋을거 같아.

```suggestion
        _state.update { it.reducer() }
```

https://medium.com/geekculture/atomic-updates-with-mutablestateflow-dc0331724405